src/cookbook/games/firestore-multiplayer.md

```diff
@@ -0,0 +1,469 @@
+    _log.fine('Received new data from Firestore (${snapshot.data()})');
+
+    final cards = snapshot.data() ?? [];
+
+    if (listEquals(cards, area.cards)) {
+      _log.fine('No change');
+    } else {
+      _log.fine('Updating local data with Firestore data ($cards)');
+      area.replaceWith(cards);
+    }
+  }
+}
+
+class FirebaseControllerException implements Exception {
+  final String message;
+
+  FirebaseControllerException(this.message);
+
+  @override
+  String toString() => 'FirebaseControllerException: $message';
+}
+```
+
+Notice the following features of this code:
+
+-   The controller's constructor takes a `BoardState`.
+    This enables the controller to manipulate the local state of the game.
+
+-   The controller subscribes to both local changes to update Firestore
+    and to remote changes to update the local state and UI.
+
+-   The fields `_areaOneRef` and `_areaTwoRef` are
+    Firebase document references.
+    They describe where the data for each area resides,
+    and how to convert between the local Dart objects (`List<PlayingCard>`)
+    and remote JSON objects (`Map<String, dynamic>`).
+    The Firestore API lets us subscribe to these references
+    with `.snapshots()`, and write to them with `.set()`.
+
+
+## 5. Use the Firestore controller
+
+1.  Open the file responsible for starting the play session:
+    `lib/play_session/play_session_screen.dart` in the case of the
+    `card` template.
+    You instantiate the Firestore controller from this file.
+
+2.  Import Firebase and the controller:
+
+    ```dart
+    import 'package:cloud_firestore/cloud_firestore.dart';
+    import '../multiplayer/firestore_controller.dart';
+    ```
+
+3.  Add a nullable field to the `_PlaySessionScreenState` class
+    to contain a controller instance:
+
+    ```dart
+    FirestoreController? _firestoreController;
+    ```
+
+4.  In the `initState()` method of the same class,
+    add code that tries to read the FirebaseFirestore instance
+    and, if successful, constructs the controller.
+    You added the `FirebaseFirestore` instance to `main.dart`
+    in the _Initialize Firestore_ step.
+
+    ```dart
+    final firestore = context.read<FirebaseFirestore?>();
+    if (firestore == null) {
+      _log.warning("Firestore instance wasn't provided. "
+          "Running without _firestoreController.");
+    } else {
+      _firestoreController = FirestoreController(
+        instance: firestore,
+        boardState: _boardState,
+      );
+    }
+    ```
+
+5.  Dispose of the controller using the `dispose()` method
+    of the same class.
+
+    ```dart
+    _firestoreController?.dispose();
+    ```
+
+
+## 6. Test the game
+
+1.  Run the game on two separate devices
+    or in 2 different windows on the same device.
+
+2.  Watch how adding a card to an area on one device
+    makes it appear on the other one.
+
+    {% comment %}
+      TBA: GIF of multiplayer working
```

Comment:
Just checking in to see if you plan to add this now or are tracking adding it later.

Reply:
This is for later. Maybe just a few days for now, but I don't want to block landing this on it, because making a good gif always takes me quite a lot of time.